### ISSUE TITLE
Within JWKS, do we expect certs, bare keys, or both?

### ISSUE THREAD

**Issue by jmandel** [smart-on-fhir/fhir-bulk-data-docs#43]
When creating a JWKS, each key can be represented in a few ways:

* `x5c` property (a PEM-encoded certificate chain)
* `x5u` property (a URL leading to a PEM-encoded certificate chain)
* "bare key" with properties like
  * `n` and `e` (modulus and exponent, for RSA keys) 
  * `crv`, `x`, and `y` (curve, x-coordinate, and y-coordinate, for EC keys) 

We want to avoid a situation where some servers expect `x5u` values and other servers expect bare keys, without a client being able to tell the difference.

From the perspective of interop, I'd like to propose that a client should always include bare key values, and a server should always be able to process bare key values.

@daliboz @kpshek @isaacvetter and others, I would love feedback. Especially

1) Do you agree we should say something about this?
2) Do you agree with making bare keys the "common denominator" for implementations?

**Comment by kpshek** [comment]
@jmandel - I imagine the decision for whether we should prescribe and/or recommend an approach has a lot to do with the support for these options in the popular open source libraries. Have you researched each of the popular libraries across the common languages to see what they support?

**Comment by jmandel** [comment]
I'll do a review of some of the most popular libraries. Please add others here if you want :)

**Comment by jmandel** [comment]
C#: https://github.com/dvsekhvalnov/jose-jwt uses .NET classes like RSACryptoServiceProvider which have pretty broad support. It looks simple to use bare keys; I'm sure there's also a way to parse certificate chains but it's less trivial to find examples.

Python: https://pyjwt.readthedocs.io/en/latest/algorithms.html?highlight=elliptic supports loading from PEM encoded keys, but not directly from cert chains or bare keys. So, takes some work to support either option.

Node.js: https://github.com/auth0/node-jsonwebtoken seems to support (require?) PEM-encoded public keys, so it would take some work to process keys in either of the formats here (i.e., cert chains or bare key) -- but https://github.com/Brightspace/node-jwk-to-pem gives you a bridge, anyway.

Bottom line: I don't see great library support for cert chains or bare keys directly; choosing a language with good openssl support at least lets you do some of this processing to create your own bridge (e.g. from public key to bare key, or from cert chain to public key, or whatever)...

**Comment by whitehatguy** [comment]
I personally would vote for bare keys, unless the goal of the specification intends to allow other "regional" variants/profiles to exist (read below).

From precursory reading, it appears the OIDC community may have "punted" on this particular question as well.  I found [a thread][1] which indicates that they expected (in the OIDC use case) that identity providers in the EU would need to leverage PKI as part of their high-assurance trust framework; this commentary leads me to believe that the choice for format in the specification was less about choice in interop mechanisms and more about allowing implementations to choose a public key format applicable to their needs.

Turning this question on its head - is this particular specification meant to be more of a _framework_, allowing varying implementations to exist in the wild that each allow participants to conform to the rules of local (perhaps government-mandated) trust profiles?  Alternatively, is it intended to be a concrete protocol that dictates the rules of interoperability?  

In the former case, I'd leave it open; in the latter case, I'd agree with the proposal to use bare keys.  Bare keys would eliminate the possibility of implementers assuming intent around the semantics of the certificate (revocation checking, checking the entirety of the certificate chain to a trusted root authority, etc.) and applying their own rules around the usage of those certificates.

[1]: https://groups.google.com/forum/#!topic/openid-connect-interop/VWRcuHM7OKw

**Comment by isaacvetter** [comment]
Hey @whitehatguy !

This spec is a concrete protocol that dictates the rules of interoperability. Note that we're even [recommending specific JWT signing algorithms](https://github.com/smart-on-fhir/fhir-bulk-data-docs/issues/18) to encourage interoperability.

**Comment by whitehatguy** [comment]
Hey @isaacvetter - given that, I'd recommend avoiding PEMs as to prevent implementations from attempting to use the certificate content for anything more than as a delivery vehicle for public key material.  I've run into similar problems in SAML development previously where implementations assume / require such content to conform to their own requirements.

**Comment by jmandel** [comment]
Thanks @whitehatguy. I'm inclined to bare keys as well; I think we should try this, recognizing that any choice is going to be a bit painful, and will benefit from a small collection of language-specific examples.

**Comment by lcmaas** [comment]
My read of section 4 of the JSON Web Key specification (RFC 7517) is that the key-type specific parameters (i.e. "bare key" parameters) must always be present in a valid JWK, regardless of whether or not x5c or x5u are included, so IMO the draft text is consistent with the JWK spec.

At the same time, I think that using X.509 certificates together with this back-end protocol may become important for future scalability, for cross-organizational queries, and as a way to validate client keys to more specific trust profiles than can be achieved by solely relying on a server's TLS certificate and its associated certificate semantics. So, I would just like to be sure that we don't inadvertently preclude the use of x5c or x5u by trust communities who want to do so.

**Comment by jmandel** [comment]
Thanks @lcmaas. We definitely won't prevent the use of certs as an optional addition.

Would you mind sharing a direct reference to the text that makes you say:

> My read of section 4 of the JSON Web Key specification (RFC 7517) is that the key-type specific parameters (i.e. "bare key" parameters) must always be present in a valid JWK, regardless of whether or not x5c or x5u are included, so IMO the draft text is consistent with the JWK spec.

I was having trouble finding support for this idea, and it'd be great to be able to cite chapter and verse!

**Comment by lcmaas** [comment]
RFC 7517 Section 4 Paragraph 2 (italics added by LCM):

   In addition to the common parameters, _each JWK will have members that
   are key type specific_.  These members represent the parameters of the
   key.  Section 6 of the JSON Web Algorithms (JWA) [JWA] specification
   defines multiple kinds of cryptographic keys and their associated
   members.

IMO "MUST" would have been better than "will" in the first sentence above, but nonetheless there is no optionality stated or implied in that paragraph or in referenced Section of the JWA spec (RFC 7518) with respect to inclusion of the type-specific members. The referenced members from Section 6 of RFC 7518 are in fact the "bare" key parameters we have been discussing in this thread, i.e. kty, EC: crv, x, (y), RSA: n,e. 

Both the x5c and x5u parameters are unconditionally labelled as optional in Sections 4.6 and 4.7 of RFC 7517. There is no text to suggest that x5c or x5u can be used as an alternative to the type-specific JWA parameters.

**Comment by jmandel** [comment]
Perfect, and thanks for the quick response!

**Comment by jmandel** [comment]
For the record: https://tools.ietf.org/html/rfc7518#section-6.2.1 and https://tools.ietf.org/html/rfc7518#section-6.3.1 in particular define the bare key parameters we've discussed.

**Comment by jmandel** [comment]
Closed by https://github.com/smart-on-fhir/fhir-bulk-data-docs/pull/46
